samples/star/coil-rule/src/main/kotlin/com/slack/circuit/sample/coil/test/FakeImageLoader.kt

```diff
@@ -1,60 +1,48 @@
 // Copyright (C) 2022 Slack Technologies, LLC
 // SPDX-License-Identifier: Apache-2.0
-package com.slack.circuit.star.ui
+package com.slack.circuit.sample.coil.test
 
-import android.graphics.Color
-import android.graphics.drawable.ColorDrawable
-import android.graphics.drawable.Drawable
-import androidx.annotation.VisibleForTesting
 import coil.ComponentRegistry
 import coil.ImageLoader
-import coil.decode.DataSource
+import coil.annotation.ExperimentalCoilApi
 import coil.disk.DiskCache
 import coil.memory.MemoryCache
 import coil.request.DefaultRequestOptions
 import coil.request.Disposable
 import coil.request.ImageRequest
 import coil.request.ImageResult
-import coil.request.SuccessResult
+import coil.test.FakeImageLoaderEngine
 import kotlinx.coroutines.CompletableDeferred
 
 /**
  * Fake Coil ImageLoader based on example found here:
  * https://coil-kt.github.io/coil/image_loaders/#testing
+ *
+ * This delegates everything to a backing [FakeImageLoaderEngine] that intercepts all calls.
  */
-@VisibleForTesting(otherwise = VisibleForTesting.NONE)
-class FakeImageLoader(private val drawable: Drawable = ColorDrawable(Color.LTGRAY)) : ImageLoader {
+@OptIn(ExperimentalCoilApi::class)
+internal class FakeImageLoader(engine: FakeImageLoaderEngine) : ImageLoader {
 
   override val defaults = DefaultRequestOptions()
-  override val components = ComponentRegistry()
+  override val components = ComponentRegistry.Builder().add(engine).build()
   override val memoryCache: MemoryCache?
     get() = null
   override val diskCache: DiskCache?
     get() = null
 
   override fun enqueue(request: ImageRequest): Disposable {
-    request.target?.onStart(request.placeholder)
-    request.target?.onSuccess(drawable)
-
     return object : Disposable {
-      override val job = CompletableDeferred(newResult(request, drawable))
+      override val job = CompletableDeferred<ImageResult>()
       override val isDisposed
         get() = true
       override fun dispose() = Unit
     }
   }
 
-  override suspend fun execute(request: ImageRequest): ImageResult = newResult(request, drawable)
-
-  private fun newResult(request: ImageRequest, drawable: Drawable): SuccessResult {
-    return SuccessResult(
-      drawable = drawable,
-      request = request,
-      dataSource = DataSource.MEMORY_CACHE
-    )
-  }
+  override suspend fun execute(request: ImageRequest): ImageResult =
+    throw AssertionError("This should never be called.")
 
-  override fun newBuilder() = throw UnsupportedOperationException()
+  override fun newBuilder() = throw AssertionError("This should never be called.")
```

Comment:
AssertionError is more appropriate for tests. `UnsupportedOperationException` is more like `TODO()` or `NotImplementedError`. It's not wrong per se, but it could be a bit ambiguous/a toe hold whereas we want to make it clear here that this should never be called in this test.